### PR TITLE
Add backing asset value check to redeem method

### PR DESCRIPTION
**Pull request by shahthepro**

### CODE REVIEW

File: contracts/contracts/vault/VaultCore.sol [hunk 1]
```diff
@@ -17,6 +17,10 @@ import { IVault } from "../interfaces/IVault.sol";
 contract VaultCore is VaultStorage {
     uint256 constant MAX_UINT = 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff;
 
+    uint256 maxSupplyDiff;
```
Comment (tomlinton) [inline contracts/contracts/vault/VaultCore.sol:20]: The storage and the event should be in VaultStorage

File: contracts/contracts/vault/VaultCore.sol [hunk 2]
```diff
@@ -611,6 +632,23 @@ contract VaultCore is VaultStorage {
         return assets[_asset].isSupported;
     }
 
+    /**
+     * @dev Returns the maximum allowable difference between
+     * total supply and backing assets' value.
+     */
+    function getMaxSupplyDiff() external view returns (uint256) {
```
Comment (tomlinton) [inline contracts/contracts/vault/VaultCore.sol:639]: We can just remove this getter and use public on the storage

File: contracts/contracts/vault/VaultCore.sol [hunk 3]
```diff
@@ -611,6 +632,23 @@ contract VaultCore is VaultStorage {
...
+        return maxSupplyDiff;
+    }
+
+    /**
+     * @dev Sets the maximum allowable difference between
+     * total supply and backing assets' value.
+     */
+    function setMaxSupplyDiff(uint256 _maxSupplyDiff) external onlyGovernor {
```
Comment (tomlinton) [inline contracts/contracts/vault/VaultCore.sol:647]: Lets move this to VaultAdmin

File: contracts/contracts/vault/VaultCore.sol [hunk 4]
```diff
@@ -158,6 +162,23 @@ contract VaultCore is VaultStorage {
...
+        if (maxSupplyDiff > 0) {
+            // Allow a max difference of maxSupplyDiff% between
+            // backing assets value and OUSD total supply
+            require(
+                (
+                    _totalSupply > _backingValue
+                        ? _totalSupply.div(_backingValue)
+                        : _backingValue.div(_totalSupply)
+                )
+                    .mul(100 ether) <= maxSupplyDiff,
```
Comment (tomlinton) [inline contracts/contracts/vault/VaultCore.sol:177]: Can we use StableMath here? Most other units we store are expressed in 18 decimal.

maxSupplyDiff should be expressed like 2e17 for 20%.

**Comment by shahthepro** [comment]
@tomlinton Have addressed your comments and added a test too.

**Pull request merged**